### PR TITLE
drivers/flash: stm32: Extend qspi-nor support to F7 series - no DMA

### PR DESCRIPTION
**Pull request by hjuul**
This PR adds qspi flash support to stm32 f7. DMA is not supported (will be added in #31405).

I have verified that `samples/drivers/spi_flash` works on boards _stm32f746g_disco_ and _stm32f769i_disco_.

### CODE REVIEW

File: boards/arm/stm32f746g_disco/stm32f746g_disco.dts [hunk 1]
```diff
@@ -119,3 +120,36 @@
...
+			#size-cells = <1>;
+
+			slot1_partition: partition@0 {
+				label = "image-1";
+				reg = <0x00000000 0x000D8000>;
+				};
+
+			storage_partition: partition@D8000 {
+				label = "storage";
+				reg = <0x000D8000 DT_SIZE_M(7)>;
```
Comment (hjuul) [inline boards/arm/stm32f746g_disco/stm32f746g_disco.dts:151]: Not sure which values I should put for the partition sizes. I just copied these from _disco_l475_iot1_. Comments appreciated.

File: boards/arm/stm32f746g_disco/stm32f746g_disco.dts [hunk 2]
```diff
@@ -119,3 +120,36 @@
...
+	pinctrl-0 = <&quadspi_clk_pb2 &quadspi_bk1_ncs_pb6
+		     &quadspi_bk1_io0_pd11 &quadspi_bk1_io1_pd12
+		     &quadspi_bk1_io2_pe2 &quadspi_bk1_io3_pd13>;
+	status = "okay";
+
+	n25q128a1: qspi-nor-flash@0 {
+		compatible = "st,stm32-qspi-nor";
+		label = "N25Q128A1";
+		reg = <0>;
+		qspi-max-frequency = <108000000>;
```
Comment (erwango) [inline boards/arm/stm32f746g_disco/stm32f746g_disco.dts:134]: This value does not wok on my end, but 107000000 is fine.
Maybe there is some tolerance to be taken into account between parts ?

File: boards/arm/stm32f746g_disco/stm32f746g_disco.dts (hunk 1 above)
Comment (erwango) [reply boards/arm/stm32f746g_disco/stm32f746g_disco.dts:151]: This board and disco have the same internal flash, so you can use the same image flash, but for storage, you can use the max amount minus image size (around 15M).
Though, to be coherent, I'd suggest adding the flash partition on internal flash as well.

File: dts/arm/st/f7/stm32f7.dtsi [hunk 3]
```diff
@@ -679,6 +679,16 @@
 			status = "disabled";
 			label = "SDMMC_1";
 		};
+		quadspi: quadspi@a0001000 {
```
Comment (erwango) [inline dts/arm/st/f7/stm32f7.dtsi:682]: Please break a line between nodes

File: boards/arm/stm32f769i_disco/stm32f769i_disco.dts [hunk 4]
```diff
@@ -104,3 +105,36 @@ arduino_serial: &usart6 {};
...
+		status = "okay";
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			slot1_partition: partition@0 {
+				label = "image-1";
+				reg = <0x00000000 0x000D8000>;
```
Comment (erwango) [inline boards/arm/stm32f769i_disco/stm32f769i_disco.dts:131]: Image size can be doubled up.

File: boards/arm/stm32f769i_disco/stm32f769i_disco.dts [hunk 5]
```diff
@@ -104,3 +105,36 @@ arduino_serial: &usart6 {};
...
+				};
+
+			storage_partition: partition@D8000 {
+				label = "storage";
+				reg = <0x000D8000 DT_SIZE_M(7)>;
```
Comment (erwango) [inline boards/arm/stm32f769i_disco/stm32f769i_disco.dts:136]: I'd set Max size - image size = 62

**Review by erwango — CHANGES_REQUESTED** [review]

File: boards/arm/stm32f746g_disco/stm32f746g_disco.dts (hunk 2 above)
Comment (hjuul) [reply boards/arm/stm32f746g_disco/stm32f746g_disco.dts]: Setting it to 107MHz results in a qspi frequency of 72MHz. I chose to change the value to 100MHz, which also ends up at 72MHz.

File: boards/arm/stm32f746g_disco/stm32f746g_disco.dts (hunk 1 above)
Comment (hjuul) [reply boards/arm/stm32f746g_disco/stm32f746g_disco.dts]: Actually, had to resize flash image because sector size is much larger for F7, so we need at least 256KiB scratch partition.

File: boards/arm/stm32f746g_disco/stm32f746g_disco.dts (hunk 2 above)
Comment (erwango) [reply boards/arm/stm32f746g_disco/stm32f746g_disco.dts]: Ok, maybe it would be better to set 72MHz directly then (Less surprise for users and avoid useless computation).

Comment (hjuul) [reply boards/arm/stm32f746g_disco/stm32f746g_disco.dts]: Ok, I can do that.

But what do you recommend I do for 769i_disco? I have set it to 70MHz because that is the most conservative spec from the data sheet. If we assume 70MHz should be used then `qspi-max-frequency` should be set to 54MHz to follow your reasoning.

But the flash can go significantly faster if more dummy cycles are allowed. Based on your knowledge of the qspi driver implementation, can you recommend a max frequency? Data sheet is here:
 https://www.mxic.com.tw/Lists/Datasheet/Attachments/7881/MX25L51245G,%203V,%20512Mb,%20v1.7.pdf

Comment (erwango) [reply boards/arm/stm32f746g_disco/stm32f746g_disco.dts]: I don't have strong opinions other than the max tested ok value

**Review by erwango — APPROVED** [review]

**Review by FRASTM — APPROVED** [review]

**Comment by erwango** [comment]
@hjuul can you rebase ?

**Comment by hjuul** [comment]
@erwango I've rebased. One of the checks are failing but I can't figure out why. Could you please have a look?

**Comment by erwango** [comment]
@hjuul Seems you're out of luck and rebased at the wrong time. 
If you mind rebasing one more time, you'll get https://github.com/zephyrproject-rtos/zephyr/pull/32435 that should fix the PWM issue

**Comment by erwango** [comment]
CI Failure: https://github.com/zephyrproject-rtos/zephyr/issues/32344

File: boards/arm/stm32f746g_disco/stm32f746g_disco.dts [hunk 6]
```diff
@@ -119,3 +120,67 @@
 		     &eth_txd0_pg13
 		     &eth_txd1_pg14>;
 };
+
+&flash0 {
```
Comment (MaureenHelm) [inline boards/arm/stm32f746g_disco/stm32f746g_disco.dts:124]: Is this related to qspi-nor or is it internal flash? It looks like something that might belong in a separate commit

Comment (hjuul) [reply boards/arm/stm32f746g_disco/stm32f746g_disco.dts:124]: This part is related to internal flash. @erwango suggested I add it for consistency with other disco boards.

Comment (gmarull) [inline boards/arm/stm32f746g_disco/stm32f746g_disco.dts:124]: flash layout changes are not related to QSPI, so they should go into a separate commit (or even a new PR if not related/required with these changes)

**Review by gmarull — CHANGES_REQUESTED** [review]

File: boards/arm/stm32f746g_disco/stm32f746g_disco.dts (hunk 6 above)
Comment (hjuul) [reply boards/arm/stm32f746g_disco/stm32f746g_disco.dts]: I've removed the flash layout for internal flash and rebased.

**Review by MaureenHelm — APPROVED** [review]

File: boards/arm/stm32f769i_disco/stm32f769i_disco.dts [hunk 7]
```diff
@@ -104,3 +105,36 @@ arduino_serial: &usart6 {};
...
+		     &quadspi_bk1_io2_pe2 &quadspi_bk1_io3_pd13>;
+	status = "okay";
+
+	mx25l51245g: qspi-nor-flash@0 {
+		compatible = "st,stm32-qspi-nor";
+		label = "MX25L51245G";
+		reg = <0>;
+		qspi-max-frequency = <72000000>;
+		/* 512 Megabits = 64 Megabytes */
+		size = <0x20000000>;
```
Comment (gmarull) [inline boards/arm/stm32f769i_disco/stm32f769i_disco.dts:121]: Better use `DT_SIZE_M` helper

File: boards/arm/stm32f746g_disco/stm32f746g_disco.dts [hunk 8]
```diff
@@ -119,3 +120,36 @@
...
+		     &quadspi_bk1_io2_pe2 &quadspi_bk1_io3_pd13>;
+	status = "okay";
+
+	n25q128a1: qspi-nor-flash@0 {
+		compatible = "st,stm32-qspi-nor";
+		label = "N25Q128A1";
+		reg = <0>;
+		qspi-max-frequency = <72000000>;
+		/* 128 Megabits = 16 Megabytes */
+		size = <0x8000000>;
```
Comment (gmarull) [inline boards/arm/stm32f746g_disco/stm32f746g_disco.dts:136]: Better use `DT_SIZE_M` helper

**Review by gmarull — CHANGES_REQUESTED** [review]

**Review by gmarull — APPROVED** [review]
